### PR TITLE
Add dao

### PR DESCRIPTION
**Pull request by jinSasaki**
Add a SQLite client.
Add evaluation and event DAOs.

- [x] SQLite client
- [x] EvaluationDAO
- [x] EvaluationDAOTests
- [x] EventDAO
- [x] EventDAOTests

ref https://github.com/ca-dp/bucketeer-android-sdk/pull/64

### CODE REVIEW

File: Bucketeer2/Sources/Internal/Database/SQLite/SQLite.swift [hunk 1]
```diff
@@ -0,0 +1,97 @@
+import Foundation
+import SQLite3
+
+final class SQLite {
```
Comment (jinSasaki) [inline Bucketeer2/Sources/Internal/Database/SQLite/SQLite.swift:4]: This is a SQLite client created not to depend on a library.
It is in reference to [SQLite.swift](https://github.com/stephencelis/SQLite.swift)

File: Bucketeer.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -30,9 +30,9 @@
...
 		00696D6D28D1D189000068E1 /* Migration1to2.swift in Sources */ = {isa = PBXBuildFile; fileRef = 00696D6C28D1D189000068E1 /* Migration1to2.swift */; };
-		00696D7128D1D258000068E1 /* CurrentEvaluationDao.swift in Sources */ = {isa = PBXBuildFile; fileRef = 00696D7028D1D258000068E1 /* CurrentEvaluationDao.swift */; };
-		00696D7328D1D295000068E1 /* CurrentEvaluationEntity.swift in Sources */ = {isa = PBXBuildFile; fileRef = 00696D7228D1D295000068E1 /* CurrentEvaluationEntity.swift */; };
-		00696D7528D1D2BB000068E1 /* CurrentEvaluationDaoImpl.swift in Sources */ = {isa = PBXBuildFile; fileRef = 00696D7428D1D2BB000068E1 /* CurrentEvaluationDaoImpl.swift */; };
+		00696D7128D1D258000068E1 /* EvaluationDao.swift in Sources */ = {isa = PBXBuildFile; fileRef = 00696D7028D1D258000068E1 /* EvaluationDao.swift */; };
+		00696D7328D1D295000068E1 /* EvaluationEntity.swift in Sources */ = {isa = PBXBuildFile; fileRef = 00696D7228D1D295000068E1 /* EvaluationEntity.swift */; };
+		00696D7528D1D2BB000068E1 /* EvaluationDaoImpl.swift in Sources */ = {isa = PBXBuildFile; fileRef = 00696D7428D1D2BB000068E1 /* EvaluationDaoImpl.swift */; };
```
Comment (jinSasaki) [inline Bucketeer.xcodeproj/project.pbxproj:34]: Renamed and modified. https://github.com/ca-dp/bucketeer-android-sdk/pull/64

File: Bucketeer2/Sources/Internal/Database/Migration/Migration.swift [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+final class Migration1to2: Migration {
+    private let db: SQLite
+
+    init(db: SQLite) {
+        self.db = db
+    }
+
+    func migration() throws {
+        let evaluationTable = SQLite.Table(entity: EvaluationEntity())
```
Comment (cre8ivejp) [inline Bucketeer2/Sources/Internal/Database/Migration/Migration.swift:13]: Don't we need to drop the table first before creating it?

Comment (jinSasaki) [reply Bucketeer2/Sources/Internal/Database/Migration/Migration.swift:19]: Added SQL for deletion, as you pointed out. Thanks! https://github.com/ca-dp/bucketeer-ios-sdk/pull/83/commits/5ccafa24e9aa5cd8b95bfd731fe7fa7495adb068

Comment (cre8ivejp) [reply Bucketeer2/Sources/Internal/Database/Migration/Migration.swift:19]: Thank you!

Would it be possible for you to add a unit test for the migration process?

E.g.
https://github.com/bucketeer-io/android-client-sdk/blob/main/bucketeer/src/test/kotlin/io/bucketeer/sdk/android/internal/database/migration/MigrationTest.kt

Comment (jinSasaki) [reply Bucketeer2/Sources/Internal/Database/Migration/Migration.swift:19]: Sorry, I forgot!
I'll add it.

Comment (jinSasaki) [reply Bucketeer2/Sources/Internal/Database/Migration/Migration.swift:19]: Added the test! ✅   https://github.com/ca-dp/bucketeer-ios-sdk/pull/83/commits/73bf5b1a2d1b9f17ed84defd3178d9cedab18d88

File: Bucketeer2Tests/EvaluationDaoTests.swift [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        try! dao.put(userId: "user1", evaluations: mocks)
+    }
+
+    override func tearDown() {
+        try! FileManager.default.removeItem(at: url)
+
+        super.tearDown()
+    }
+
+    func testUpsert() throws {
```
Comment (cre8ivejp) [inline Bucketeer2Tests/EvaluationDaoTests.swift:67]: It says upsert but is only inserting a new row. Please rename it.

File: Bucketeer2Tests/EvaluationDaoTests.swift [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+        super.setUp()
+
+        let db = try! SQLite(path: path, logger: nil)
+
+        let evaluationTable = SQLite.Table(entity: EvaluationEntity())
+        let evaluationSql = evaluationTable.sqlToCreate()
+        try! db.exec(query: evaluationSql)
+
+        let dao = EvaluationDaoImpl(db: db)
+        try! dao.put(userId: "user1", evaluations: mocks)
```
Comment (cre8ivejp) [inline Bucketeer2Tests/EvaluationDaoTests.swift:58]: Reading the test cases is easier and safe if you don't add data in the setup step.
Doing this doesn't cover cases if the interactor adds a single item or multiple items.

Would you mind covering other cases like in the android tests, please?
The same for EventDaoImplTest.

### Evaluation
https://github.com/bucketeer-io/android-client-sdk/blob/main/bucketeer/src/test/kotlin/io/bucketeer/sdk/android/internal/evaluation/db/EvaluationDaoImplTest.kt

### Event
https://github.com/bucketeer-io/android-client-sdk/blob/main/bucketeer/src/test/kotlin/io/bucketeer/sdk/android/internal/event/db/EventDaoImplTest.kt

File: Bucketeer2Tests/EvaluationDaoTests.swift [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+        XCTAssertEqual(evaluations.count, 1)
+        guard let evaluation = evaluations.first else {
+            return
+        }
+        XCTAssertEqual(evaluation.id, "evaluation3")
+        XCTAssertEqual(evaluation.variation.id, "variation3")
+        XCTAssertEqual(evaluation.reason.rule_id, "rule3")
+    }
+
+    func testDelete() throws {
```
Comment (cre8ivejp) [inline Bucketeer2Tests/EvaluationDaoTests.swift:99]: The name says delete but is deleting and inserting. Please rename it.

File: Bucketeer2Tests/EvaluationDaoTests.swift (hunk 5 above)
Comment (jinSasaki) [reply Bucketeer2Tests/EvaluationDaoTests.swift]: Covered other cases https://github.com/ca-dp/bucketeer-ios-sdk/pull/83/commits/919ab813b4c163e17ae2b678af4d30128f75bd90 with simple mock data https://github.com/ca-dp/bucketeer-ios-sdk/pull/83/commits/c42c76f80da9e40fb60ef3f7dffa5707a009da60 🙆‍♀️

File: Bucketeer2Tests/EvaluationDaoTests.swift (hunk 6 above)
Comment (jinSasaki) [reply Bucketeer2Tests/EvaluationDaoTests.swift]: Renamed and covered other cases https://github.com/ca-dp/bucketeer-ios-sdk/pull/83/commits/919ab813b4c163e17ae2b678af4d30128f75bd90

File: Bucketeer2Tests/EvaluationDaoTests.swift [hunk 7]
```diff
@@ -64,68 +22,138 @@ final class EvaluationDaoTests: XCTestCase {
...
+        try dao.put(userId: "user1", evaluations: [.mock1, .mock2])
+
+        let evaluations = try dao.get(userId: "user1")
+        XCTAssertEqual(evaluations.count, 2)
+        XCTAssertEqual(evaluations[0], Evaluation.mock1)
+        XCTAssertEqual(evaluations[1], Evaluation.mock2)
     }
 
-    func testDelete() throws {
+    func testDeleteAllAndInsertToInsert() throws {
```
[the remark this answers is not in the document]
Comment (jinSasaki) [reply Bucketeer2Tests/EvaluationDaoTests.swift]: Thanks for your review!
Renamed test case https://github.com/ca-dp/bucketeer-ios-sdk/pull/83/commits/a79b1f6b9491d8132647eb789aab2ec6e971e544

**Review by cre8ivejp — APPROVED** [review]
Great!
Thank you!

**Comment by jinSasaki** [comment]
Thank you for your review! I'll merge.

**Pull request merged**